Content.Server/TapeRecorder/TapeRecorderSystem.cs

```diff
@@ -0,0 +1,345 @@
+
+                //stop player if tape is at end
+                if (tapeRecorder.Enabled && tapeRecorder.InsertedTape.TimeStamp >= tapeRecorder.InsertedTape.TapeMaxTime && tapeRecorder.CurrentMode != TapeRecorderState.Rewind)
+                {
+                    StopTape(tapeRecorder);
+                    ChangeMode(tapeRecorder, TapeRecorderState.Rewind); // go into rewind mode once at end of tape
+                }
+
+                //Handle tape playback
+                if (tapeRecorder.Enabled && tapeRecorder.CurrentMode == TapeRecorderState.Play)
+                {
+
+                    if (tapeRecorder.CurrentMessageIndex >= tapeRecorder.InsertedTape.RecordedMessages.Count || tapeRecorder.InsertedTape.TimeStamp >= tapeRecorder.InsertedTape.TapeMaxTime)
+                    {
+                        StopTape(tapeRecorder);
+                        ChangeMode(tapeRecorder, TapeRecorderState.Rewind); // go into rewind mode once we reached the end of recorded data
+                        return;
+                    }
+
+                    //send the current message once our tapes timestamp passes the message timestamp
+                    if (tapeRecorder.InsertedTape.TimeStamp > tapeRecorder.InsertedTape.RecordedMessages[tapeRecorder.CurrentMessageIndex].MessageTimeStamp)
+                    {
+                        _chat.TrySendInGameICMessage(tapeRecorder.Owner, tapeRecorder.InsertedTape.RecordedMessages[tapeRecorder.CurrentMessageIndex].Message, InGameICChatType.Speak, false);
+                        tapeRecorder.CurrentMessageIndex++;
+                    }
+
+                    tapeRecorder.InsertedTape.TimeStamp += frameTime;
+                }
+
+                //Tape rewinding. fast-forwarding later (maybe)
+                if (!tapeRecorder.Enabled || tapeRecorder.CurrentMode != TapeRecorderState.Rewind)
+                    continue;
+
+                tapeRecorder.InsertedTape.TimeStamp -= frameTime * tapeRecorder.RewindSpeed;
+
+                if (tapeRecorder.InsertedTape.TimeStamp <= 0) //stop rewinding when we get to the beginning of the tape
+                {
+                    tapeRecorder.InsertedTape.TimeStamp = 0;
+                    tapeRecorder.CurrentMessageIndex = 0;
+                    StopTape(tapeRecorder);
+                    ChangeMode(tapeRecorder, TapeRecorderState.Play); // go into play mode once finished rewinding
+                    return;
+                }
+
+            }
+        }
+
+        private void OnItemSlotChanged(EntityUid uid, TapeRecorderComponent component, ref ItemSlotChangedEvent args)
+        {
+            StopTape(component);
+
+            if (!_containerSystem.TryGetContainer(uid, "cassette_tape", out var container) || container is not ContainerSlot slot)
+                return;
+
+            if (!TryComp<CassetteTapeComponent>(slot.ContainedEntity, out var cassetteTapeComponent))
+            {
+                component.InsertedTape = null;
```

Comment:
Might want to eject it too if it's something that wouldn't be able to pass the whitelist anymore?